test/src/lib.rs

```diff
@@ -1 +1,432 @@
+                d.read_element::<asn1::BigInt>()?; // serial-number
+                d.read_element::<asn1::Sequence>()?; // signature
+                d.read_element::<asn1::Sequence>()?; // name
+                d.read_element::<asn1::Sequence>()?; // validity
+                d.read_element::<asn1::Sequence>()?; // subject
+                d.read_element::<asn1::Sequence>()?; // subjectPublicKeyInfo
+                d.read_optional_implicit_element::<asn1::BitString>(1)?; // issuerUniqueID
+                d.read_optional_implicit_element::<asn1::BitString>(2)?; // subjectUniqueId
+                let result = d.read_explicit_element::<asn1::Sequence>(3)?.parse(|d| {
+                    let mut result = None;
+                    while !d.is_empty() {
+                        let found_result = d.read_element::<asn1::Sequence>()?.parse(|d| {
+                            let item_oid = d.read_element::<asn1::ObjectIdentifier>()?;
+                            d.read_element::<Option<bool>>()?; // critical
+                            let value = d.read_element::<&[u8]>()?;
+                            if &item_oid == oid {
+                                Ok(Some(value))
+                            } else {
+                                Ok(None)
+                            }
+                        })?;
+                        if let Some(found_result) = found_result {
+                            if result.is_some() {
+                                // The extension was found more than once
+                                return Err(asn1::ParseError::new(asn1::ParseErrorKind::ExtraData));
+                            }
+                            result = Some(found_result);
+                        }
+                    }
+                    Ok(result)
+                })?;
+                Ok(result)
+            })?;
+            d.read_element::<asn1::Sequence>()?; // signatureAlgorithm
+            d.read_element::<asn1::BitString>()?; // signatureValue
+            Ok(result)
+        })
+    })
+}
+
+#[derive(Eq, PartialEq)]
+pub struct DiceFwid {
+    pub hash_alg: asn1::ObjectIdentifier,
+    pub digest: Vec<u8>,
+}
+impl std::fmt::Debug for DiceFwid {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DiceFwid")
+            .field("hash_alg", &format!("{}", &self.hash_alg))
+            .field("digest", &format!("{:02x?}", self.digest))
+            .finish()
+    }
+}
+
+#[derive(Debug, Default, Eq, PartialEq)]
+pub struct DiceTcbInfo {
+    pub vendor: Option<String>,
+    pub model: Option<String>,
+    pub version: Option<String>,
+    pub svn: Option<u32>,
+    pub layer: Option<u32>,
+    pub index: Option<u32>,
+    pub fwids: Vec<DiceFwid>,
+    pub flags: Option<u32>,
+    pub vendor_info: Option<Vec<u8>>,
+    pub ty: Option<Vec<u8>>,
+}
+
+impl DiceTcbInfo {
+    fn parse(d: &mut asn1::Parser) -> Result<Self, asn1::ParseError> {
```

Comment:
It would probably be easier to use the asn1 crate's parsing trait. See https://github.com/chipsalliance/caliptra-dpe/blob/7ad2c990bf9e06a1374bb2a02a8289761ee8e97f/dpe/src/x509.rs#LL965C1-L965C1

Reply:
I started down that path originally, but didn't like that I couldn't use ownership types (string, vec, etc) as the field types.

Reply:
Could just implement Asn1Writable and Asn1Readable for the types you want to use, right? And can use the asn1 functions to implement them, so I'd think they would be small implementations. But I could be wrong.

If it's not too much work, I think it will make things a lot easier/more readable to be able to use the annotations for optional (implicit/explicit) fields.

Reply:
Rust doesn't let you implement an external trait (like AsnReadable) for an external type (like String, Vec). Either the trait or type need to be in my crate.